examples/pytorch/graphsage/multi_gpu_node_classification.py

```diff
@@ -1,182 +1,164 @@
+
+def run(proc_id, nprocs, devices, g, dist_data, mode):
+    # find corresponding device for my rank
+    device = devices[proc_id]
+    torch.cuda.set_device(device)
+    # initialize process group and unpack data for sub-processes
+    dist.init_process_group(backend="nccl", init_method='tcp://127.0.0.1:12345', world_size=nprocs, rank=proc_id)
+    out_size, train_idx, val_idx, test_idx = dist_data
+    train_idx = train_idx.to(device)
+    val_idx = val_idx.to(device)
+    g = g.to(device if mode == 'puregpu' else 'cpu')
+    # create RGCN model (distributed)
+    in_size = g.ndata['feat'].shape[1]
+    model = SAGE(in_size, 256, out_size).to(device)
+    model = DistributedDataParallel(model, device_ids=[device], output_device=device)
+    # training + testing
+    use_uva = (mode == 'mixed')
+    train(proc_id, nprocs, device, g, train_idx, val_idx, model, use_uva)
+    layerwise_infer(proc_id, device, g, test_idx, model, use_uva, batch_size=2**16)    
+    # cleanup process group
+    dist.destroy_process_group()
 
 if __name__ == '__main__':
-    dataset = DglNodePropPredDataset('ogbn-products')
-    graph, labels = dataset[0]
-    graph.ndata['label'] = labels
-    graph.create_formats_()     # must be called before mp.spawn().
-    split_idx = dataset.get_idx_split()
-    num_classes = dataset.num_classes
-    # use all available GPUs
-    n_procs = torch.cuda.device_count()
-
-    # Tested with mp.spawn and fork.  Both worked and got 4s per epoch with 4 GPUs
-    # and 3.86s per epoch with 8 GPUs on p2.8x, compared to 5.2s from official examples.
-    import torch.multiprocessing as mp
-    mp.spawn(train, args=(n_procs, graph, num_classes, split_idx), nprocs=n_procs)
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--mode", default='mixed', choices=['mixed', 'puregpu'],
+                        help="Training mode. 'mixed' for CPU-GPU mixed training, "
+                             "'puregpu' for pure-GPU training.")
+    parser.add_argument("--gpu", type=str, default='0',
+                           help="GPU(s) in use. Can be a list of gpu ids for multi-gpu training,"
+                                " e.g., 0,1,2,3.")
+    args = parser.parse_args()
+    devices = list(map(int, args.gpu.split(',')))
+    nprocs = len(devices)
+    assert torch.cuda.is_available(), f"Must have GPUs to enable multi-gpu training."
+    print(f'Training in {args.mode} mode using {nprocs} GPU(s)')
+    
+    # load and preprocess dataset
+    print('Loading data')
+    dataset = AsNodePredDataset(DglNodePropPredDataset('ogbn-products'))
+    g = dataset[0]
+    # avoid creating certain graph formats in each sub-process to save momory
+    g.create_formats_()
+    # thread limiting to avoid resource competition
+    os.environ['OMP_NUM_THREADS'] = str(mp.cpu_count() // 2 // nprocs)    
+    dist_data = dataset.num_classes, dataset.train_idx, dataset.val_idx, dataset.test_idx
```

Comment:
Not sure if `dist_data` is better than `data`

Reply:
Right. changed `dist_data` --> `data`